File: web/src/logic.ts
import type { BudgetState, DeletedExpense, Expense, MonthSummary, RecurringExpense } from './types'

export function todayIso(): string {
  // Use local date (not UTC) to match Python's date.today() and user expectations.
  const d = new Date()
  const y = d.getFullYear()
  const m = String(d.getMonth() + 1).padStart(2, '0')
  const day = String(d.getDate()).padStart(2, '0')
  return `${y}-${m}-${day}`
}

export function monthKeyFromIso(isoDate: string): string {
  return isoDate.slice(0, 7)
}

export function currentMonthKey(): string {
  return monthKeyFromIso(todayIso())
}

export function prevMonthKey(monthKey: string): string {
  const year = Number(monthKey.slice(0, 4))
  const month = Number(monthKey.slice(5, 7))
  if (month === 1) return `${String(year - 1).padStart(4, '0')}-12`
  return `${String(year).padStart(4, '0')}-${String(month - 1).padStart(2, '0')}`
}

export function computeRemainingDaysInMonth(fromIso: string): number {
  // IMPORTANT: do not use millisecond diffs across calendar dates.
  // DST transitions (e.g. March) can make a "day" be 23/25 hours and
  // produce off-by-one results. Use pure calendar math instead.
  const [, , d] = fromIso.split('-').map(Number)
  const daysInMonth = computeDaysInMonth(fromIso)
  const remaining = daysInMonth - d + 1
  return Math.max(1, remaining)
}

export function computeDaysInMonth(fromIso: string): number {
  const [y, m] = fromIso.split('-').map(Number)
  // Use UTC so local DST offset changes cannot affect the result.
  // JS months are 0-based; day 0 gives the last day of the previous month.
  const lastDayUtc = new Date(Date.UTC(y, m, 0))
  return lastDayUtc.getUTCDate()
}

export function sumExpenses(expenses: Expense[]): number {
  return expenses.reduce((acc, e) => acc + (Number.isFinite(e.amount) ? e.amount : 0), 0)
}

export function expensesForDate(expenses: Expense[], isoDate: string): Expense[] {
  return expenses.filter((e) => e.date === isoDate)
}

export function expensesForMonth(expenses: Expense[], monthKey: string): Expense[] {
  const prefix = `${monthKey}-`
  return expenses.filter((e) => typeof e.date === 'string' && e.date.startsWith(prefix))
}

export function computeSpendPerDay(baseAmount: number, daysInMonth: number): number {
  const days = daysInMonth <= 0 ? 1 : daysInMonth
  return baseAmount / days
}

export function computeNoRewardSpendPerDay(
  baseAmount: number,
  daysInMonth: number,
  remainingDaysInclToday: number,
  overspendDebt: number
): number {
  // "No reward" model:
  // - Daily target is static: base / days_in_month.
  // - Underspending does NOT increase future per-day allowance.
  // - Overspending creates a debt that reduces future per-day allowance.
  const baseline = computeSpendPerDay(baseAmount, daysInMonth)
  const days = remainingDaysInclToday <= 0 ? 1 : remainingDaysInclToday
  const debt = Math.max(0, overspendDebt)
  const penalty = debt / days
  return Math.max(0, baseline - penalty)
}

export function computeOverspendDebt(
  expenses: Expense[],
  monthKey: string,
  nowIso: string,
  baselinePerDay: number,
  includeToday = false
): number {
  // Overspend debt is the sum of (spent_that_day - baseline)+ for prior days.
  // This intentionally does NOT let underspending "bank" credit.
  // includeToday=true is used to compute the projected allowance for tomorrow.
  const monthExpenses = expensesForMonth(expenses, monthKey)
  const totalsByDate: Record<string, number> = {}
  for (const e of monthExpenses) {
    if (typeof e.date !== 'string') continue
    // includeToday=false: exclude today + future (e.date >= nowIso)
    // includeToday=true: exclude only future (e.date > nowIso)
    if (includeToday ? e.date > nowIso : e.date >= nowIso) continue
    const amt = Number.isFinite(e.amount) ? e.amount : 0
    totalsByDate[e.date] = (totalsByDate[e.date] ?? 0) + amt
  }

  let debt = 0
  for (const total of Object.values(totalsByDate)) {
    const overspend = total - baselinePerDay
    if (overspend > 0) debt += overspend
  }
  return debt
}

export function computeRecurringTotal(recurring: RecurringExpense[]): number {
  return recurring.reduce((sum, r) => sum + (Number.isFinite(r.amount) ? r.amount : 0), 0)
}

// Count consecutive days from yesterday backwards where daily spend was <= baseline.
// Days with no logged expenses count as $0 (under budget).
export function computeStreak(
  expenses: Expense[],
  monthKey: string,
  nowIso: string,
  baselinePerDay: number
): number {
  const monthExpenses = expensesForMonth(expenses, monthKey)
  const totalsByDate: Record<string, number> = {}
  for (const e of monthExpenses) {
    if (typeof e.date !== 'string' || e.date >= nowIso) continue
    const amt = Number.isFinite(e.amount) ? e.amount : 0
    totalsByDate[e.date] = (totalsByDate[e.date] ?? 0) + amt
  }

  const [y, m, d] = nowIso.split('-').map(Number)
  let streak = 0
  let day = d - 1 // start at yesterday

  while (day >= 1) {
    const isoDate = `${String(y).padStart(4, '0')}-${String(m).padStart(2, '0')}-${String(day).padStart(2, '0')}`
    const spent = totalsByDate[isoDate] ?? 0
    if (spent > baselinePerDay) break
    streak++
    day--
  }

  return streak
}

// Projected daily allowance for tomorrow if the user stops spending today.
// Includes today's expenses in the debt, uses remainingDays - 1 as the future window.
export function computeProjectedAllowance(
  expenses: Expense[],
  base: number,
  daysInMonth: number,
  remainingDaysInclToday: number,
  monthKey: string,
  nowIso: string
): number {
  const baselinePerDay = computeSpendPerDay(base, daysInMonth)
  const debtIncludingToday = computeOverspendDebt(expenses, monthKey, nowIso, baselinePerDay, true)
  const remainingAfterToday = Math.max(1, remainingDaysInclToday - 1)
  return computeNoRewardSpendPerDay(base, daysInMonth, remainingAfterToday, debtIncludingToday)
}

export function normalizeState(raw: unknown): BudgetState {
  const obj = (raw && typeof raw === 'object') ? (raw as any) : {}
  const base_amount = Number(obj.base_amount)
  const expenses = Array.isArray(obj.expenses) ? obj.expenses : []
  const parsedExpenses: Expense[] = []
  for (const item of expenses) {
    if (!item || typeof item !== 'object') continue
    const date = String((item as any).date ?? '')
    const amount = Number((item as any).amount)
    const note = (item as any).note
    if (!date || !Number.isFinite(amount)) continue
    parsedExpenses.push({ date, amount, note: typeof note === 'string' ? note : undefined })
  }

  const monthly_bases_raw = obj.monthly_bases
  const monthly_bases: Record<string, number> = {}
  if (monthly_bases_raw && typeof monthly_bases_raw === 'object') {
    for (const [k, v] of Object.entries(monthly_bases_raw as Record<string, unknown>)) {
      const n = Number(v)
      if (typeof k === 'string' && Number.isFinite(n)) monthly_bases[k] = n
    }
  }

  const recurring: RecurringExpense[] = []
  if (Array.isArray(obj.recurring)) {
    for (const item of obj.recurring) {
      if (!item || typeof item !== 'object') continue
      const r = item as any
      const id = typeof r.id === 'string' && r.id ? r.id : String(Date.now() + Math.random())
      const amount = Number(r.amount)
      const day = Math.round(Number(r.day))
      if (!Number.isFinite(amount) || amount <= 0) continue
      if (!Number.isFinite(day) || day < 1 || day > 31) continue
      recurring.push({ id, amount, day, note: typeof r.note === 'string' ? r.note : '' })
    }
  }

  const recently_deleted: DeletedExpense[] = []
  if (Array.isArray(obj.recently_deleted)) {
    for (const item of obj.recently_deleted) {
      if (!item || typeof item !== 'object') continue
      const d = item as any
      const deleted_at = typeof d.deleted_at === 'string' ? d.deleted_at : ''
      if (!deleted_at) continue
      const exp = d.expense
      if (!exp || typeof exp !== 'object') continue
      const date = String((exp as any).date ?? '')
      const amount = Number((exp as any).amount)
      const note = (exp as any).note
      if (!date || !Number.isFinite(amount)) continue
      recently_deleted.push({
        expense: { date, amount, note: typeof note === 'string' ? note : undefined },
        deleted_at
      })
    }
  }

  const monthly_summaries: Record<string, MonthSummary> = {}
  if (obj.monthly_summaries && typeof obj.monthly_summaries === 'object') {
    for (const [k, v] of Object.entries(obj.monthly_summaries as Record<string, unknown>)) {
      if (typeof k !== 'string' || !v || typeof v !== 'object') continue
      const sv = v as any
      const base = Number(sv.base)
      const spent = Number(sv.spent)
      const saved = Number(sv.saved)
      if (Number.isFinite(base) && Number.isFinite(spent) && Number.isFinite(saved)) {
        monthly_summaries[k] = { base, spent, saved }
      }
    }
  }

  const themeRaw = obj.theme
  const theme: 'light' | 'dark' = themeRaw === 'dark' ? 'dark' : 'light'

  return {
    base_amount: Number.isFinite(base_amount) ? base_amount : 0,
    expenses: parsedExpenses,
    monthly_bases,
    last_rollover_month: typeof obj.last_rollover_month === 'string' ? obj.last_rollover_month : '',
    last_month_key: typeof obj.last_month_key === 'string' ? obj.last_month_key : '',
    last_month_saved: Number.isFinite(Number(obj.last_month_saved)) ? Number(obj.last_month_saved) : 0,
    recurring,
    recently_deleted,
    monthly_summaries,
    theme
  }
}

export function rolloverMonthIfNeeded(state: BudgetState, nowIso: string): { changed: boolean } {
  const currentKey = monthKeyFromIso(nowIso)
  if ((state.last_rollover_month ?? '') === currentKey) return { changed: false }

  const prevKey = prevMonthKey(currentKey)
  const prevSpent = sumExpenses(expensesForMonth(state.expenses, prevKey))
  const prevBase = state.monthly_bases?.[prevKey] ?? state.base_amount
  const prevSaved = prevBase - prevSpent

  state.last_rollover_month = currentKey
  state.last_month_key = prevKey
  state.last_month_saved = prevSaved

  if (!state.monthly_summaries) state.monthly_summaries = {}
  state.monthly_summaries[prevKey] = { base: prevBase, spent: prevSpent, saved: prevSaved }

  // Keep at most 12 months of summaries
  const keys = Object.keys(state.monthly_summaries).sort()
  if (keys.length > 12) {
    for (const k of keys.slice(0, keys.length - 12)) {
      delete state.monthly_summaries[k]
    }
  }

  return { changed: true }
}

File: src/budget_app/logic.py
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import date, timedelta


@dataclass
class Expense:
    date: str  # ISO: YYYY-MM-DD
    amount: float
    note: str = ""


@dataclass
class RecurringExpense:
    id: str
    amount: float
    day: int   # day of month (1–31) when the expense occurs
    note: str = ""


@dataclass
class DeletedExpense:
    date: str
    amount: float
    note: str
    deleted_at: str  # ISO datetime string, used for 24h pruning


@dataclass
class MonthSummary:
    base: float
    spent: float
    saved: float


@dataclass
class BudgetState:
    base_amount: float = 0.0
    expenses: list[Expense] = field(default_factory=list)

    # Per-month base amounts ("planned to spend") captured when the user hits Save.
    # Keys are "YYYY-MM".
    monthly_bases: dict[str, float] = field(default_factory=dict)

    # Month rollover bookkeeping.
    last_rollover_month: str = ""
    last_month_key: str = ""
    last_month_saved: float = 0.0

    # Recurring expenses pre-deducted from the monthly base.
    recurring: list[RecurringExpense] = field(default_factory=list)

    # Recently deleted expenses buffer for undo (pruned after 24h).
    recently_deleted: list[DeletedExpense] = field(default_factory=list)

    # Compact per-month summaries for historical view (up to 12 months).
    monthly_summaries: dict[str, MonthSummary] = field(default_factory=dict)

    # UI theme preference.
    theme: str = "light"


# ── Date helpers ─────────────────────────────────────────────────────────────

def compute_remaining_days_in_month(today: date | None = None) -> int:
    """Remaining days in current month including today."""
    if today is None:
        today = date.today()
    if today.month == 12:
        first_next = date(today.year + 1, 1, 1)
    else:
        first_next = date(today.year, today.month + 1, 1)
    last_this = first_next - timedelta(days=1)
    return max(1, (last_this - today).days + 1)


def compute_days_in_month(today: date | None = None) -> int:
    """Total number of days in the current month."""
    if today is None:
        today = date.today()
    if today.month == 12:
        first_next = date(today.year + 1, 1, 1)
    else:
        first_next = date(today.year, today.month + 1, 1)
    first_this = date(today.year, today.month, 1)
    return max(1, (first_next - first_this).days)


# ── Expense helpers ───────────────────────────────────────────────────────────

def expenses_for_date(expenses: list[Expense], target: date) -> list[Expense]:
    target_iso = target.isoformat()
    return [e for e in expenses if e.date == target_iso]


def expenses_for_month(expenses: list[Expense], year: int, month: int) -> list[Expense]:
    prefix = f"{year:04d}-{month:02d}-"
    return [e for e in expenses if isinstance(e.date, str) and e.date.startswith(prefix)]


def sum_expenses(expenses: list[Expense]) -> float:
    total = 0.0
    for e in expenses:
        try:
            total += float(e.amount)
        except (TypeError, ValueError):
            continue
    return total


# ── Budget calculations ───────────────────────────────────────────────────────

def compute_recurring_total(recurring: list[RecurringExpense]) -> float:
    """Sum of all recurring expense amounts."""
    total = 0.0
    for r in recurring:
        try:
            total += float(r.amount)
        except (TypeError, ValueError):
            continue
    return total


def compute_spend_per_day(base_amount: float, days_in_month: int) -> float:
    if days_in_month <= 0:
        days_in_month = 1
    return base_amount / float(days_in_month)


def compute_overspend_debt_for_month(
    expenses: list[Expense],
    *,
    year: int,
    month: int,
    through_date_exclusive: date,
    baseline_per_day: float,
    include_through_date: bool = False,
) -> float:
    """Sum overspending debt for prior days in the month.

    include_through_date=True includes through_date_exclusive itself (used for
    projected allowance calculation).
    """
    prefix = f"{year:04d}-{month:02d}-"
    cutoff_iso = through_date_exclusive.isoformat()
    totals_by_date: dict[str, float] = {}

    for e in expenses:
        if not isinstance(e.date, str) or not e.date.startswith(prefix):
            continue
        if include_through_date:
            if e.date > cutoff_iso:
                continue
        else:
            if e.date >= cutoff_iso:
                continue
        try:
            amt = float(e.amount)
        except (TypeError, ValueError):
            continue
        totals_by_date[e.date] = totals_by_date.get(e.date, 0.0) + amt

    debt = 0.0
    for total in totals_by_date.values():
        overspend = total - float(baseline_per_day)
        if overspend > 0:
            debt += overspend
    return debt


def compute_conservative_carryover_per_day(
    base_amount: float,
    *,
    days_in_month: int,
    remaining_days_incl_today: int,
    overspend_debt: float,
) -> float:
    """Compute per-day allowance using a "no reward" model."""
    if days_in_month <= 0:
        days_in_month = 1
    if remaining_days_incl_today <= 0:
        remaining_days_incl_today = 1
    baseline = compute_spend_per_day(base_amount, days_in_month)
    debt = max(0.0, float(overspend_debt))
    penalty = debt / float(remaining_days_incl_today)
    return max(0.0, baseline - penalty)


def compute_projected_allowance(
    expenses: list[Expense],
    base_amount: float,
    *,
    year: int,
    month: int,
    today: date,
    days_in_month: int,
    remaining_days_incl_today: int,
) -> float:
    """Allowance for tomorrow if the user stops spending today.

    Includes today's expenses in the debt and uses remaining_days - 1 as the
    future window.
    """
    baseline_per_day = compute_spend_per_day(base_amount, days_in_month)
    debt_including_today = compute_overspend_debt_for_month(
        expenses,
        year=year,
        month=month,
        through_date_exclusive=today,
        baseline_per_day=baseline_per_day,
        include_through_date=True,
    )
    remaining_after_today = max(1, remaining_days_incl_today - 1)
    return compute_conservative_carryover_per_day(
        base_amount,
        days_in_month=days_in_month,
        remaining_days_incl_today=remaining_after_today,
        overspend_debt=debt_including_today,
    )


def compute_streak(
    expenses: list[Expense],
    year: int,
    month: int,
    today: date,
    baseline_per_day: float,
) -> int:
    """Count consecutive days from yesterday backwards where spend <= baseline."""
    prefix = f"{year:04d}-{month:02d}-"
    today_iso = today.isoformat()
    totals_by_date: dict[str, float] = {}

    for e in expenses:
        if not isinstance(e.date, str) or not e.date.startswith(prefix):
            continue
        if e.date >= today_iso:
            continue
        try:
            amt = float(e.amount)
        except (TypeError, ValueError):
            continue
        totals_by_date[e.date] = totals_by_date.get(e.date, 0.0) + amt

    streak = 0
    day = today.day - 1  # start at yesterday
    while day >= 1:
        iso = f"{year:04d}-{month:02d}-{day:02d}"
        spent = totals_by_date.get(iso, 0.0)
        if spent > baseline_per_day:
            break
        streak += 1
        day -= 1
    return streak
